EKF/airspeed_fusion.cpp

```diff
@@ -46,152 +47,137 @@
+	Vector24f Kfusion; // Kalman gain vector
 
-	// Calculate the predicted airspeed
-	const float v_tas_pred = sqrtf((ve - vwe) * (ve - vwe) + (vn - vwn) * (vn - vwn) + vd * vd);
+	const float &vn = _state.vel(0); // Velocity in north direction
+	const float &ve = _state.vel(1); // Velocity in east direction
+	const float &vd = _state.vel(2); // Velocity in downwards direction
+	const float &vwn = _state.wind_vel(0); // Wind speed in north direction
+	const float &vwe = _state.wind_vel(1); // Wind speed in east direction
 
 	// Variance for true airspeed measurement - (m/sec)^2
 	const float R_TAS = sq(math::constrain(_params.eas_noise, 0.5f, 5.0f) *
 			    math::constrain(_airspeed_sample_delayed.eas2tas, 0.9f, 10.0f));
 
-	// Perform fusion of True Airspeed measurement
-	if (v_tas_pred > 1.0f) {
-		// determine if we need the sideslip fusion to correct states other than wind
-		const bool update_wind_only = !_is_wind_dead_reckoning;
-
-		// intermediate variable from algebraic optimisation
-		float SH_TAS[3];
-		SH_TAS[0] = 1.0f/v_tas_pred;
-		SH_TAS[1] = (SH_TAS[0]*(2.0f*ve - 2.0f*vwe))*0.5f;
-		SH_TAS[2] = (SH_TAS[0]*(2.0f*vn - 2.0f*vwn))*0.5f;
-
-		// Observation Jacobian
-		Vector24f H_TAS;
-		H_TAS(4) = SH_TAS[2];
-		H_TAS(5) = SH_TAS[1];
-		H_TAS(6) = vd*SH_TAS[0];
-		H_TAS(22) = -SH_TAS[2];
-		H_TAS(23) = -SH_TAS[1];
-
-		_airspeed_innov_var = (R_TAS + SH_TAS[2]*(P(4,4)*SH_TAS[2] + P(5,4)*SH_TAS[1] - P(22,4)*SH_TAS[2] - P(23,4)*SH_TAS[1] + P(6,4)*vd*SH_TAS[0]) + SH_TAS[1]*(P(4,5)*SH_TAS[2] + P(5,5)*SH_TAS[1] - P(22,5)*SH_TAS[2] - P(23,5)*SH_TAS[1] + P(6,5)*vd*SH_TAS[0]) - SH_TAS[2]*(P(4,22)*SH_TAS[2] + P(5,22)*SH_TAS[1] - P(22,22)*SH_TAS[2] - P(23,22)*SH_TAS[1] + P(6,22)*vd*SH_TAS[0]) - SH_TAS[1]*(P(4,23)*SH_TAS[2] + P(5,23)*SH_TAS[1] - P(22,23)*SH_TAS[2] - P(23,23)*SH_TAS[1] + P(6,23)*vd*SH_TAS[0]) + vd*SH_TAS[0]*(P(4,6)*SH_TAS[2] + P(5,6)*SH_TAS[1] - P(22,6)*SH_TAS[2] - P(23,6)*SH_TAS[1] + P(6,6)*vd*SH_TAS[0]));
-
-		if (_airspeed_innov_var >= R_TAS) { // Check for badly conditioned calculation
-			_fault_status.flags.bad_airspeed = false;
-
-		} else { // Reset the estimator covariance matrix
-			_fault_status.flags.bad_airspeed = true;
-
-			// if we are getting aiding from other sources, warn and reset the wind states and covariances only
-			const char* action_string = nullptr;
-			if (update_wind_only) {
-				resetWindStates();
-				resetWindCovariance();
-				action_string = "wind";
-
-			} else {
-				initialiseCovariance();
-				_state.wind_vel.setZero();
-				action_string = "full";
-			}
-			ECL_ERR("airspeed badly conditioned - %s covariance reset", action_string);
-
-			return;
-		}
+	// determine if we need the sideslip fusion to correct states other than wind
+	const bool update_wind_only = !_is_wind_dead_reckoning;
+
+	// Intermediate variables
+	const float HK0 = vn - vwn;
+	const float HK1 = ve - vwe;
+	const float HK2 = ecl::powf(HK0, 2) + ecl::powf(HK1, 2) + ecl::powf(vd, 2);
```

Comment:
The if-statement would be a bit clearer when the variable `v_tas_speed` would be used